Recorder upgrades (queue, FLAC, fonts, range sort)

TheAMM opened the pull request:
Here's a bunch of goodies for the recorder feature (`dump-cache` that is, `--record-file` can rot but does gain here too).
Includes fixes for some slight human errors, but also support for cache dumping FLAC streams (previously broken), streams with very long GOPs (also previously broken) and attaching embedded fonts when outputting to matroska (quite useful with ASS subtitles).

The packet queue (long GOPs) commit fixes #8278.

sfan5 commented on common/av_common.c:
```diff
@@ -80,12 +80,16 @@ AVCodecParameters *mp_codec_params_to_av(struct mp_codec_params *c)
-        avp->extradata =
-            av_mallocz(c->extradata_size + AV_INPUT_BUFFER_PADDING_SIZE);
+        // The internal matroska demuxer keeps the flac header - lavf doesn't
+        int skip = strncmp(c->extradata, "fLaC\0\0\0\0", 8) ? 0 : 8;
```
Checking the codec_id/tag in addition would be safer and a good idea here IMO.

Also needs to be `memcmp`.

TheAMM replied:
Yeah this was bad and it can go think about itself in the void. (The `\0\0\0\0` was also broken (and added after initial tests): the expected last byte is 0x22 (34) for the STREAMINFO block size.)

sfan5 commented on common/recorder.c:
```diff
@@ -153,6 +156,37 @@ struct mp_recorder *mp_recorder_create(struct mpv_global *global,
+                attachment->data_size + AV_INPUT_BUFFER_PADDING_SIZE
+            );
+            if (!a_stream->codecpar->extradata) {
+                MP_ERR(priv, "Can't allocate memory for one of "
```
I don't think mpv usually prints log messages for allocation failures, not sure tho.

TheAMM replied:
Gone.

TheAMM commented:
Replaced 7b2311edbe23f94104c5807948bc4c8824b19c21 (av_common FLAC extradata fix) with more proper a9fb11cd31741581ee02a10be3078123d6cfe630 (demux_mkv FLAC extradata fix) as per IRC stuff, and removed the allocation error log from recorder.c font muxing.

sfan5 commented on demux/demux_mkv.c:
```diff
@@ -1743,15 +1743,18 @@ static int demux_mkv_open_audio(demuxer_t *demuxer, mkv_track_t *track)
+            // only the initial STREAMINFO after parsing the channel map,
+            // so let's do the same for compability (say, recorder.c muxing)
+            extradata += 8;
+            extradata_len = 34; // FLAC_STREAMINFO_SIZE
```
This is not safe without checking how many bytes are actually present.
Why not `extradata_len -= 8;`?

TheAMM replied:
> Why not extradata_len -= 8;?

Because we have no need for the extended extradata, only the FLAC STREAMINFO is parsed. The remaining is usually just the channel mapping (which we/ffmpeg generate when muxing), and other(?) possible non-playback metadata.

> This is not safe

Fair, I'll go with a `8 + 34` in the if.

sfan5 replied:
Ah, so it would break if the rest of the data is kept?

sfan5 reviewed (approved):
Fine by me anyway.

TheAMM commented:
Force-pushed the updated range sort commit as discussed on IRC.

TheAMM commented:
Forced again to replace the matroska demuxer patch a9fb11cd31741581ee02a10be3078123d6cfe630 with an av_common FLAC extradata fix 68f8c5b636e1adaca83f6b614c8fa6ce48a88a13.

jeeb reviewed (approved):
LGTM.

The reason for doing the extradata filtering on the muxing side of things is because the interface between a demuxer and the following component for decoder initialization data AKA extradata (usually decoder, but can be other things like the recorder) is the decoder initialization data specified in the container (just like avc or hevc or av1 configuration record is passed on to decoders).

I have only taken a limited look at the packet queue removal bits, but since @sfan5 has noted that he has taken a look at that, I am OK with that.
